homeassistant/components/sensor/history_stats.py

```diff
@@ -0,0 +1,303 @@
+            return
+
+        # Get the first state
+        last_state = history.get_state(start, self._entity_id)
+        last_state = (last_state is not None and
+                      last_state == self._entity_state)
+        last_time = start_timestamp
+        elapsed = 0
+
+        # Make calculations
+        for item in history_list.get(self._entity_id):
+            current_state = item.state == self._entity_state
+            current_time = item.last_changed.timestamp()
+
+            if last_state:
+                elapsed += current_time - last_time
+
+            last_state = current_state
+            last_time = current_time
+
+        # Save value in hours
+        self.value = elapsed / 3600
+
+    def update_period(self):
+        """Parse the templates and store a timestamp tuple in _period."""
+        start = None
+        end = None
+        duration = None
+
+        try:
+            if self._start is not None:
+                start = math.floor(float(self._start.async_render()))
+            if self._end is not None:
+                end = math.floor(float(self._end.async_render()))
+            if self._duration is not None:
+                duration = math.floor(float(self._duration.async_render()))
+        except TemplateError as ex:
+            HistoryStatsHelper.handle_template_exception(ex)
+        except ValueError:
+            _LOGGER.error('Template value cannot be converted to timestamp ')
+
+        # Calculate start or end using the duration
+        start = end - duration if start is None else start
+        end = start + duration if end is None else end
+
+        self._period = start, end
+
+
+class HistoryStatsHelper:
+    """Static methods to make the HistoryStatsSensor code lighter."""
+
+    @staticmethod
+    def parse_time_expr(expression):
+        """Replace time expressions with functions accepted by templates."""
+        regex = r"(_NOW_(\.replace\([a-z]+=[0-9]+\))*)"
+        replacement = r"as_timestamp(\1)"
+
+        return re.sub(
+            regex, replacement, expression.replace("_THIS_DAY_", "_TODAY_")
+            .replace("_ONE_WEEK_", "604800")
+            .replace("_ONE_DAY_", "86400")
+            .replace("_ONE_HOUR_", "3600")
+            .replace("_ONE_MINUTE_", "60")
+            .replace("_THIS_YEAR_", "_THIS_MONTH_.replace(month=1)")
+            .replace("_THIS_MONTH_", "_TODAY_.replace(day=1)")
+            .replace("_THIS_WEEK_", "_TODAY_ - now().weekday() * 86400")
+            .replace("_TODAY_", "_THIS_HOUR_.replace(hour=0)")
+            .replace("_THIS_HOUR_", "_THIS_MINUTE_.replace(minute=0)")
+            .replace("_THIS_MINUTE_", "_NOW_.replace(second=0)")) \
+            .replace("_NOW_", "now()")
+
+    @staticmethod
+    def pretty_datetime(timestamp):
+        """Format a timestamp to a formatted datetime in the local timezone."""
+        return dt_util.as_local(dt_util.utc_from_timestamp(timestamp)) \
+            .strftime('%Y/%m/%d %H:%M:%S')
```

Comment:
This would introduce an addition time format. We are trying to use ISO 8601 everything. Also, Europeans use a different format (at least I do anyway) ;-)